### PR TITLE
Improve T5 docs 

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR aims to clarify and explain some of the magic that's happening when using T5 for training/inference. It includes:
- explaining that the model automatically creates the `decoder_input_ids` based on the `labels` (a lot of people were still confused by this, see e.g. #11977, #13213 
- added code examples, to show a basic forward pass that includes the fact that padding token ids of the labels should be replaced by -100 (at least, for PyTorch, I see that for FLAX one uses the `decoder_attention_mask` to skip padding tokens), and code examples for inference (both batched/not batched)
- adding info about T5's variants, including T5v1.1, mT5 and byT5, with links to their docs.
- additional tips & tricks, based on what I found on the forum (learning rate, training on TPU, etc).

In addition, I've added T5v1.1 to the main README as well making it have its own documentation page, and some more info to the mT5 docs.

### CODE REVIEW

File: docs/source/model_doc/t5.rst [hunk 1]
```diff
@@ -13,9 +13,6 @@
 T5
 -----------------------------------------------------------------------------------------------------------------------
 
-**DISCLAIMER:** This model is still a work in progress, if you see something strange, file a `Github Issue
-<https://github.com/huggingface/transformers/issues/new?assignees=&labels=&template=bug-report.md&title>`__.
-
```
Comment (NielsRogge) [inline docs/source/model_doc/t5.rst:18]: Don't think T5 is still a work in progress, so this can be removed.

File: docs/source/model_doc/t5.rst [hunk 2]
```diff
@@ -72,34 +88,173 @@ token. T5 can be trained / fine-tuned both in a supervised and unsupervised fash
...
+        labels = torch.tensor(labels)
+
+        # forward pass
+        loss = model(input_ids=input_ids, attention_mask=attention_mask, labels=labels).loss
+
+Additional training tips:
+
+- T5 models need a slightly higher learning rate than the default one set in the :obj:`Trainer`. Typically, 1e-4 and
+  3e-4 work well for almost all problems (classification, summarization, translation, question answering, question
+  generation).
```
Comment (NielsRogge) [inline docs/source/model_doc/t5.rst:185]: Not sure if this is true, cc @patrickvonplaten @patil-suraj.

Also wonder, whether T5 works well with AdamW or whether one should use AdaFactor.

**Review by patrickvonplaten — APPROVED** [review]
This is really great work @NielsRogge !!! Thanks a lot for looking into it and making the docs much nicer. I especially like that you make it very clear in the beginning how T5v1_1, MT5 and ByT5 are related to the original T5.

I left a couple of nits that could maybe improve / extend the explanation a bit, but overall this looks good to me!

File: docs/source/model_doc/t5.rst [hunk 3]
```diff
@@ -72,34 +97,194 @@ token. T5 can be trained / fine-tuned both in a supervised and unsupervised fash
...
+        input_sequence_1 = "Welcome to NYC"
+        output_sequence_1 = "Bienvenue à NYC"
+
+        input_sequence_2 = "HuggingFace is a company"
+        output_sequence_2 = "HuggingFace est une entreprise"
+
+        # encode the inputs
+        task_prefix = "translate English to French: "
+        input_sequences = [input_sequence_1, input_sequence_2]
+        encoding = tokenizer([task_prefix + sequence for sequence in input_sequences], padding='longest', max_length=max_source_length, truncation=True, return_tensors="pt")
```
Comment (sgugger) [inline docs/source/model_doc/t5.rst:179]: This line should be split in several lines with the black syntax formatting. It won't be readable in the documentation otherwise.

File: docs/source/model_doc/t5.rst [hunk 4]
```diff
@@ -72,34 +97,194 @@ token. T5 can be trained / fine-tuned both in a supervised and unsupervised fash
...
+        input_ids, attention_mask = encoding.input_ids, encoding.attention_mask
+
+        # encode the targets
+        target_encoding = tokenizer([output_sequence_1, output_sequence_2], padding='longest', max_length=max_target_length, truncation=True)
+        labels = target_encoding.input_ids
+
+        # replace padding token id's of the labels by -100
+        labels = [
+                   [(label if label != tokenizer.pad_token_id else -100) for label in labels_example] for labels_example in labels
+        ]
```
Comment (sgugger) [inline docs/source/model_doc/t5.rst:189]: Same here.

**Review by sgugger — APPROVED** [review]
Thanks for digging into this!

File: docs/source/model_doc/t5.rst (hunk 2 above)
Comment (patil-suraj) [reply docs/source/model_doc/t5.rst:206]: In my T5 fine-tuning experiments for most of the tasks using higher lr than the default 5e-5 with Adam worked better. But I'm not sure if this should apply to every task.
